lib/src/collection/lru_map.dart

```diff
@@ -0,0 +1,286 @@
+   * order is based off of [other.forEach] implementation - the first item would
+   * be the LRU, and the last item, the MRU.
+   *
+   * Because [maximumSize] is still enforced, this means if [other.length] is
+   * > [maximumSize], not all of the items in [other] may exist at the end of the
+   * operation.
+   */
+  @override
+  void addAll(Map<K, V> other) => other.forEach((k, v) => this[k] = v);
+
+  @override
+  void clear() {
+    _entries.clear();
+    _head = _tail = null;
+  }
+
+  @override
+  bool containsKey(K key) => _entries.containsKey(key);
+
+  @override
+  bool containsValue(V value) => values.contains(value);
+
+  /**
+   * Applies [action] to each key-value pair of the map in order of MRU to LRU.
+   *
+   * Calling `action` must not add or remove keys from the map.
+   */
+  @override
+  void forEach(void action(K key, V value)) {
+    var head = _head;
+    while (head != null) {
+      action(head.key, head.value);
+      head = head.next;
+    }
+  }
+
+  @override
+  int get length => _entries.length;
+
+  @override
+  bool get isEmpty => _entries.isEmpty;
+
+  @override
+  bool get isNotEmpty => _entries.isNotEmpty;
+
+  /**
+   * The keys of [this] - in order of MRU to LRU.
+   *
+   * The returned iterable does *not* have efficient `length` or `contains`.
+   */
+  @override
+  Iterable<K> get keys => new _LinkedEntryIterator(_head).map((e) => e.key);
+
+  /**
+   * The values of [this] - in order of MRU to LRU.
+   *
+   * The returned iterable does *not* have efficient `length` or `contains`.
+   */
+  @override
+  Iterable<V> get values => new _LinkedEntryIterator(_head).map((e) => e.value);
+
+  @override
+  int get maximumSize => _maximumSize;
+
+  @override
+  void set maximumSize(int maximumSize) {
+    if (maximumSize == null) throw new ArgumentError.notNull('maximumSize');
+    while (length > maximumSize) {
+      _removeLru();
+    }
+    _maximumSize = maximumSize;
+  }
+
+  /**
+   * Look up the value of [key], or add a new value if it isn't there.
+   *
+   * Returns the value associated to [key], if there is one, and promotes the
+   * key-value pair to the MRU position.
+   *
+   * Otherwise calls [ifAbsent] to get a new value, associates [key] to that
+   * [value], and then returns the new [value], with the key-value pair in the
+   * MRU position. If this causes [length] to exceed [maximumSize], then the
+   * LRU position is removed.
+   */
+  @override
+  V putIfAbsent(K key, V ifAbsent()) {
+    final entry =  _entries.putIfAbsent(
+        key, () => _createEntry(key, ifAbsent()));
+    _promoteEntry(entry);
+    return entry.value;
+  }
+
+  /**
+   * Get the value for a [key] in the [Map]. The [key] will be promoted to the
+   * 'Most Recently Used' position. *NOTE*: Calling [] inside an iteration over
+   * keys/values is currently unsupported; use [keys] or [values] if you
+   * need information about entries without modifying their position.
+   */
+  @override
+  V operator[] (K key) {
```

Comment:
https://www.dartlang.org/articles/style-guide/#dont-include-a-space-between-the-declared-name-of-a-method-operator-or-setter-and-its-parameter-list